### ISSUE TITLE
Error "stream is reset: NO_ERROR" when uploading big blob

### ISSUE THREAD

**Issue by ezze** [edeckers/react-native-blob-courier#173]
Not sure I have an issue with `react-native-blob-courier` directly but decided to report it with custom template in hope of getting any help. I am ready to provide more info if required.

I am uploading binary SQLite database to a server like this:

```typescript
const url = `${apiUrl}api/v1/client-db/merge`;
const method = 'POST';
const options: BlobUploadInput = {
  url,
  method,
  headers: {
    Authorization: `Bearer ${token}`
  },
  mimeType: 'multipart/form-data; boundary=WebAppBoundary',
  absoluteFilePath: databasePath,
  onProgress: (e: BlobProgressEvent) => {
    onProgress(e.written, e.total);
  },
  progressIntervalMilliseconds: 100,
  returnResponse: true
};

const { response } = await BlobCourier.uploadBlob(options);
```

And there are two cases I faced yet:

1. Database file is transferred successfully (file is ~140 Mb).
2. The following error is thrown immediately on calling `uploadBlob` (file is ~250 Mb):

    > [Error: stream was reset: NO_ERROR]

I googled and found [the similar issue on StackOverflow](https://stackoverflow.com/q/58958213/506695). Seems it's related to `okhttp3` which, I guess, is used by `react-native-blob-courier` under the hood. I also found there that the problem may be related to proxy configuration. We have Java backend behind `nginx` on our dev server so we will check it first. But what's embarrasses me is that I didn't get a response from `uploadBlob` so it might be that the error is thrown just before making any request to the server.

I am not familar with React Native modules development so my question is: how can I debug this issue? What's going wrong there and are there any options available to log http requests? Can `okhttp3` be configured somehow if the issue is related to its default configuration?

My environment is Android SDK Platform 29.

**Comment by edeckers** [comment]
tl; dr: I investigated the problem and the issue lies with your configuration of Nginx, indeed. You'll have to change the default `1MB` value of `client_max_body_size` to `0` in `/etc/nginx/nginx.conf` (https://github.com/joltup/rn-fetch-blob/issues/82).

--

Since you asked how to go about debugging this situation, I'll illustrate the way I usually go about this below:

# Try something that you know should work
First thing I do is to try something I know should work, using the setup you described. So I created a Pixel 3a AVD with SDK 29 in Android Studio and then ran the [Example app](https://github.com/edeckers/react-native-blob-courier/tree/develop/example) on this device. I clicked the "Download" button and "Upload" button, no problemo.

The default file size I use in the example app is 5MB though, whereas in your bug report you mentioned ~250MB, so I changed the file url to a larger file of 200MB which is the test file closest to your filesize on this site https://testfiledownload.com/. Pushed the buttons again and once more no issue at all. Good!

# Verify you receive the request you expect
Since the Example app uploads the file to `file.io` and I have absolutely no idea who runs this site let alone have access to its logs, I need another way to verify. I like to use `netcat` for that, like so `nc -l 8000` (I'm running Ubuntu), this starts a TCP socket on `0.0.0.0:8000`. Then I changed the `file.io` url in the example app to the local ip address (mind you, it is a common mistake to use `127.0.0.1` which won't work here, since that is the loopback ip of your phone) of my machine that is running the mentioned `netcat`. Ran the "download/upload"-routine once more, and I received the uploaded data just fine. Great!

# Bring nginx into the mix
After I verified that the original example app works and verified the correct upload request is sent by the app, I added the next complexity to the mix, Nginx. What I did is add a reverse proxy to `127.0.0.1:8000`, which means that I should be able to send HTTP requests to `http://<machine-ip>/` which are then passed on to the `netcat` TCP socket I mentioned earlier.

When I would open `http://<machine-ip/` through my browser or through `curl http://<machine-ip>` I received a perfectly fine HTTP GET request. But once I changed the upload url in the Example app to `http://<machine-ip>` and ran the "download/upload"-routine I would receive ... nothing. Just like you experienced. When I opened the Nginx error log though, I would see messages corresponding with my requests, reading something along the lines of `(...) client intended to send too large body: 10485964 bytes (...)`.

# Test the suggested solution
So then I applied the solution suggested [here](https://github.com/joltup/rn-fetch-blob/issues/82#issuecomment-820905621), ran the "download/upload"-routine and _BAM_, success! :)

**Comment by edeckers** [comment]
Also, I would like to add that I was never able to reproduce the exact error you're reporting, just the same behavior you're describing for large uploads. Where did you see the message you posted? In `Logcat`, a warning message popup, or a JavaScript exception?

And what version of RN are you using? I just recalled that the `okhttp`-version was bumped to version `4` in [RN0.65](https://reactnative.dev/blog/2021/08/17/version-065)

**Comment by ezze** [comment]
@edeckers Thanks a lot for such detailed explaination. Pointing out `netcat` usage for debugging is very useful, I will have it in my mind next time (I'm also using Ubuntu, by the way :)).

As for the question, we really had issues with nginx configuration so everything works for us now. We use React Native 0.64.2, and I saw the mentioned error in React Native error popup just after I call `uploadBlob`. But I'm sure I've caught and logged it by `console.error` on my own.

**Comment by edeckers** [comment]
Thank you for reporting back and I'm glad the `netcat` suggestion was useful!

As for the error: I've been trying to reproduce it with your setup, but unfortunately I wasn't able to. This irks me a _little bit_, but not enough to spend multiple hours on it, so I'm going to drop it until it rears its head again.
